

K-mers, based on seqeuncing data, can be used to estimate:
 * sequencing coverage [Genomescope](https://github.com/tbenavi1/genomescope2.0)
 * heterozygosity [Genomescope](https://github.com/tbenavi1/genomescope2.0)
 * repeat content [Genomescope](https://github.com/tbenavi1/genomescope2.0)
 * biases (e.g. contamination) [mash screen](https://github.com/marbl/Mash)

A brief introduction can be found here: [UCDavis genome assembly workshop](https://ucdavis-bioinformatics-training.github.io/2020-Genome_Assembly_Workshop/kmers/kmers)
In this workshop the following K-mer counters and tools will be used:
 * FastK
 * GeneScopeFK 
 * HiSim
 * meryl
 * merqury 
 * Genomescope 

 All the tools will be available within a docker container (ToDo). Exenmplary we will use sequencing data from:
 * The human T2T project: 
    * PacBio HiFi [SRR11292120](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR11292120&display=metadata), [SRR11292121](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR11292121&display=metadata), [SRR11292122](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR11292122&display=metadata), [SRR11292123](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR11292123&display=metadata)
    * Illimina 10X Genomics linked-reads [10x](https://github.com/marbl/CHM13/blob/master/Sequencing_data.md#10x-genomics-data)
    * Illumina HiC reads [hic](https://github.com/marbl/CHM13/blob/master/Sequencing_data.md#hi-c-data) 
    * Illumina reads (PCR-free) [SRR1997411](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR1997411&display=metadata), [SRR3189741](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR3189741&display=metadata), [SRR3189742](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR3189742&display=metadata), [SRR3189743](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR3189743&display=metadata)
 * 


